backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/logic/service/EdcContractMappingService.java

```diff
@@ -20,40 +20,150 @@
     }
 
-    public EdcContractMapping create(EdcContractMapping edcContractMapping) {
-        if (repository.findById(edcContractMapping.getPartnerBpnl()).isPresent()) {
-            return null;
+    private ContractMapping getOrCreateContractMapping(Partner partner, SubmodelType type) {
+        GeneralContractMappingRepository<? extends ContractMapping> repository = getContractMappingRepository(type);
+        ContractMapping entity = repository.findById(partner.getBpnl()).orElse(null);
+        if (entity == null) {
+            try {
+                Constructor<? extends ContractMapping> constructor = repository.getType().getConstructor();
+                entity = constructor.newInstance();
+                entity.setPartnerBpnl(partner.getBpnl());
+
+                if (type == SubmodelType.DTR) {
+                    entity.setPartnerDspUrl(partner.getEdcUrl());
+                }
+            } catch (Exception e) {
+                log.error("Error in GetOrCreateContractMapping for partner " + partner.getBpnl() +
+                    " and type " + type, e);
+                return null;
+            }
         }
-        return repository.save(edcContractMapping);
+        return entity;
     }
 
-    public EdcContractMapping update(EdcContractMapping edcContractMapping) {
-        if (!repository.findById(edcContractMapping.getPartnerBpnl()).isPresent()) {
-            return null;
+    public void saveContractMapping(Partner partner, SubmodelType type, ContractMapping contractMapping) {
+        if (!partner.getBpnl().equals(contractMapping.getPartnerBpnl())) {
+            throw new IllegalArgumentException("Mismatched partner BPNL");
+        }
+        GeneralContractMappingRepository<? extends ContractMapping> repository = getContractMappingRepository(type);
+
+        if (type == SubmodelType.DTR) {
+            // DTR always uses the default EDC URL from the Partner entity.
+            contractMapping.setPartnerDspUrl(partner.getEdcUrl());
+        }
+
+        if (contractMapping.getContractId() == null && contractMapping.getAssetId() == null && contractMapping.getPartnerBpnl() == null) {
+            throw new IllegalArgumentException("Missing contract data, cannot save: \n" + contractMapping);
         }
-        return repository.save(edcContractMapping);
+        repository.checkedSave(contractMapping);
     }
 
-    public void delete (String partnerBpnl) {
-        repository.deleteById(partnerBpnl);
+    public void invalidateContractMapping(Partner partner, SubmodelType type) {
+        GeneralContractMappingRepository<? extends ContractMapping> repository = getContractMappingRepository(type);
+        repository.deleteById(partner.getBpnl());
     }
 
-    public void delete(EdcContractMapping edcContractMapping) {
-        delete(edcContractMapping.getPartnerBpnl());
+
+
+    public void saveHrefMapping(Partner partner, SubmodelType type, String key, String value) {
```

Comment:
That would not work for all cases. 

A materialNumberCx is used as the mapping key in case of the ItemStock ContractMapping. 

But for instance, in the case of the PartType Information Submodel we are using the partnerMaterialNumber as mapping key.